### PR TITLE
Add Log Hyperbolic Cosine Function

### PR DESCRIPTION
**Pull request by kartikdutt18**
Hi everyone, I have implemented Log Hyperbolic Cosine Function. It can be referenced from [here](https://openreview.net/pdf?id=rkglvsC9Ym). Kindly suggest the changes that I need to make.
Thanks.

### CODE REVIEW

**Comment by kartikdutt18** [comment]
I will update History.md in the next commit. Thanks.

File: src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The Log-Hyperbolic-Cosine loss function is often used for to improve
+ * variational auto encoder. The Log-Hyperbolic-Cosine loss function
+ * taken log of hyperbolic cosine of difference between true values and
+ * predicted values.
```
Comment (birm) [inline src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp]: Minor readability suggestions: 
- replace "used for to" with "used to".
- Having "The Log-Hyperbolic-Cosine loss function" start two consecutive sentences doesn't read well; try a pronoun like "it" or "this function"
- instead of "taken" try "is the"

I'll do a more full readability lookover later.

Comment (kartikdutt18) [reply src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp]: Hi @birm, I will make the changes. Thanks for pointing them out.

Comment (kartikdutt18) [reply src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp]: Hi, I have made the changes. Thanks.

File: src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
+/**
+ * @file log_cosh_loss.hpp
+ * @author Kartik Dutt
+ *
+ * Definition of the Log-Hyperbolic-Cosine loss function.
+ *
+ * For more information, see the following paper.
+ * 
+ * @code
+ * article{
```
Comment (birm) [inline src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp]: If you have one, please add a permalink to this paper. 

Also, and much less importantly, it looks like most articles cited in this context have a bibtex shorthand intro (see [here for an example](https://github.com/mlpack/mlpack/blob/928aee24c75f4227b5c1702aeb0acdc9aaf486e0/src/mlpack/methods/ann/layer/elu.hpp#L84))

**Comment by birm** [comment]
This seems quite useful, but I got a little confused looking at the literature.

Keras has an implementation it calls logcosh which predates the paper you seem to be referring to and has the hyper-parameter you call `a` fixed to 2. (See the associated keras commit [here](https://github.com/tensorflow/tensorflow/commit/f1d4d18a622a3bb030ebc92b60763fe35a755bd5))

So, my questions are:
Do we have a more established or foundational paper to reference?
Is `a` a commonly used hyperparameter? Does it have a more descriptive name?

**Comment by kartikdutt18** [comment]
Hi @birm, in reference to this [paper](https://openreview.net/pdf?id=rkglvsC9Ym), `a` is a not fixed. TensorFlow and Keras both have implemented this loss function with fixed however `a` is useful for smoothening the curve as seen in the paper.
In the paper, they have only mentioned `a` as a hyper parameter.
I tried finding other papers for the same but efforts yielded no results.

File: src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp (hunk 2 above)
Comment (kartikdutt18) [reply src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp]: I will add the url, however the authors are anonymous, Should I add 

> Anonymous2019

 ?

**Comment by birm** [comment]
The code looks good to me. I'm not sure that logcosh _needs_ a citation since it's not new, with the possible exception of the hyperparameter `a`. If you think the citation is useful or necessary, I'm not sure what's good practice for citation of double blinded papers.

**Comment by kartikdutt18** [comment]
So I'll remove the citation and explain the use of the hyper-parameter a bit more clearly. I hope that's okay.

**Comment by kartikdutt18** [comment]
Hi @birm, Thank you for all the help.

File: src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+class LogCoshLoss
+{
+ public:
+  /**
+   * Create the Log-Hyperbolic-Cosine object with the specified
+   * parameters.
+   *
+   * @param a A double type value for smoothening loss function.
+   *          It must be positive real number, Sharpness of loss
+   *          function is directly propotional to `a`.
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp]: Just a is fine, I think ` is a markdown thing.

File: src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+   *          It can also act as a scaling factor hence making
+   *          the loss function more sensitive to small losses
+   *          around the origin.
+   *          Default value = 1.0
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp]: That could come right after the end of the line above, just to save some space.

File: src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp (hunk 3 above)
Comment (kartikdutt18) [reply src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp]: Hi @zoq, I have made the changes for the same. Thanks for pointing them out.

File: src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+>
+class LogCoshLoss
+{
+ public:
+  /**
+   * Create the Log-Hyperbolic-Cosine object with the specified
+   * parameters.
+   *
+   * @param a A double type value for smoothening loss function.
+   *          It must be positive real number, Sharpness of loss
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp]: `must be a positive`

File: src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+   *          function is directly propotional to a. It can also
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp]: `proportional`

File: src/mlpack/methods/ann/loss_functions/log_cosh_loss_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+#include "log_cosh_loss.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+LogCoshLoss<InputDataType, OutputDataType>::LogCoshLoss(const double a) :
+    a(a)
+{
+  Log::Assert(a > 0, "Hyper-Parameter a must be positive");
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/log_cosh_loss_impl.hpp]: Not sure, should we make it more clear that a is a parameter, `Hyper-Parameter 'a' must be positive.`?

Comment (kartikdutt18) [reply src/mlpack/methods/ann/loss_functions/log_cosh_loss_impl.hpp]: Agreed that makes more sense. Thanks.

File: src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp (hunk 6 above)
Comment (kartikdutt18) [reply src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp]: Terribly sorry for the typo. I will fix it.

Comment (zoq) [reply src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp]: No worries.

File: src/mlpack/tests/loss_functions_test.cpp [hunk 8]
```diff
@@ -396,4 +397,43 @@ BOOST_AUTO_TEST_CASE(DiceLossTest)
...
+  for (double el : output)
+  {
+    // For input = target we should get 0.0 everywhere.
+    BOOST_REQUIRE_CLOSE(el, 0.0, 1e-5);
+  }
+
+  BOOST_REQUIRE_EQUAL(output.n_rows, input.n_rows);
+  BOOST_REQUIRE_EQUAL(output.n_cols, input.n_cols);
+
+  // Test the Forward function. Loss should be 0.546621
```
Comment (zoq) [inline src/mlpack/tests/loss_functions_test.cpp]: One more style issue, missing stop at the end.

Comment (kartikdutt18) [reply src/mlpack/tests/loss_functions_test.cpp]: Made the changes for the same. I think I didn't add a stop here because it ended with decimal number. Sorry about that.

**Comment by birm** [comment]
@kartikdutt18 @zoq Everything looking ok to merge?

**Comment by kartikdutt18** [comment]
It is ready from my side. I would be more than happy to make incorporate any suggestions. Thanks.

**Comment by kartikdutt18** [comment]
Hmm, The tests have passed but GitHub still shows that they are in progress. I am not sure but I think there was a mention of this on the IRC.

**Comment by kartikdutt18** [comment]
Thank you everyone for all help. :)
